src/ppm/decoder.rs

```diff
@@ -0,0 +1,104 @@
+use std::io::Read;
+use std::io::BufReader;
+
+use color::{ColorType};
+use image::{DecodingResult, ImageDecoder, ImageResult, ImageError};
+
+/// PPM decoder
+pub struct PPMDecoder<R> {
+    reader: BufReader<R>,
+    width: u32,
+    height: u32,
+    maxwhite: u32,
+}
+
+impl<R: Read> PPMDecoder<R> {
+    /// Create a new decoder that decodes from the stream ```r```
+    pub fn new(r: R) -> ImageResult<PPMDecoder<R>> {
+        let mut buf = BufReader::new(r);
+        try!(PPMDecoder::read_next_string(&mut buf)); // Skip P6
+        let width = try!(PPMDecoder::read_next_u32(&mut buf));
+        let height = try!(PPMDecoder::read_next_u32(&mut buf));
+        let maxwhite = try!(PPMDecoder::read_next_u32(&mut buf));
+        Ok(PPMDecoder {
+            reader: buf,
+            width: width,
+            height: height,
+            maxwhite: maxwhite,
+        })
+    }
+
+  fn read_next_string(reader: &mut BufReader<R>) -> ImageResult<String> {
+      let mut bytes = Vec::new();
```

Comment:
Seems a shame to allocate for each field, while you could read the whole header into a `Vec` in one go and only then parse. May not be worth the engineering effort though.